client/src/unifyfs.c

```
@@ -740,15 +740,21 @@ off_t unifyfs_fid_global_size(int fid)
 {
     /* get meta data for this file */
     unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(fid);
-    return meta->global_size;
+    if (NULL != meta) {
+        return meta->global_size;
+    }
+    return (off_t)-1;
```

I'd replace with:
```C
assert(meta)
return meta->global_size
```
If `meta` is NULL then it's a hard bug.

---

We can't assume the fid that is passed is valid. My opinion for libraries linked into client applications is that they should never assert(). Rather, they should always return with an error.

---

I would say that in general, we should be checking the fid in the higher level functions that the user calls (like `fstat()`) and returning the error there.  That way we can return the error as early as possible.  You could add a comment above the unifyfs_fid_*_size() functions saying  "we assume fid has been verified" and keep the assert in for sanity.  That way we're not unnecessarily re-checking fid at every level of the code.  

That said, I'm ok with the way you have it.

---

I agree that the higher level functions should verify the fid, and fid checking at multiple levels is redundant. If we want to treat the `unifyfs_fid_xxx_size()` functions as simple getters, my suggestion would be to pass in the meta pointer rather than the fid.